src/storage/txn/commands/flashback_to_version_read_phase.rs

```diff
@@ -59,41 +77,63 @@ impl<S: Snapshot> ReadCommand<S> for FlashbackToVersionReadPhase {
             (key_locks, has_remain_locks) = key_locks_result?;
         }
         // Scan the writes.
-        let mut key_writes = Vec::with_capacity(0);
+        let mut keys = Vec::with_capacity(0);
         let mut has_remain_writes = false;
         // The batch is not full, we can still read.
         if self.next_write_key.is_some() && key_locks.len() < FLASHBACK_BATCH_SIZE {
+            let cur_key = RefCell::new(None);
             let key_writes_result = reader.scan_writes(
                 self.next_write_key.as_ref(),
                 self.end_key.as_ref(),
-                // To flashback `CF_WRITE` and `CF_DEFAULT`, we need to delete all keys whose
-                // commit_ts is greater than the specified version.
-                |key| key.decode_ts().unwrap() > self.version,
+                // To flashback the data, we need to get all the latest writes first by scanning
+                // every unique key in `CF_WRITE`.
+                move |key| {
+                    if let Ok(truncated_key) = key.clone().truncate_ts() {
+                        let mut cur_key_rc = cur_key.borrow_mut();
+                        if cur_key_rc.is_some() && truncated_key == *cur_key_rc.as_ref().unwrap() {
+                            return false;
+                        }
+                        *cur_key_rc = Some(truncated_key);
+                        return true;
+                    }
+                    false
+                },
                 FLASHBACK_BATCH_SIZE - key_locks.len(),
-                Some(self.version),
+                None,
             );
             statistics.add(&reader.statistics);
+            // Truncate the timestamps of all the keys.
+            let key_writes;
             (key_writes, has_remain_writes) = key_writes_result?;
+            for (key, _) in key_writes {
+                if key.decode_ts()? >= self.commit_ts {
+                    return Err(Error::from(ErrorInner::InvalidTxnTso {
+                        start_ts: self.start_ts,
+                        commit_ts: self.commit_ts,
+                    }));
+                }
+                keys.push(key.truncate_ts()?);
```

Comment:
if the latest write of one key precedes `self.version`, no need to write it again